src/cpp/shared/adventure/hero.cpp

```diff
@@ -157,44 +157,45 @@ void hero::TakeArtifact(int art) {
 }
 
 int hero::CalcMobility() {
-   int points;
-   int MOVEMENT_POINTS_TERM_CREATURE_MAX = 1500;
-   int PLAYER_FIRST = 0;
-   int PLAYER_LAST = 5;
-   
-   float gfSSLogisticsMod[] = { 1.0,  1.1,  1.2,  1.3 };
-   
-   mapCell* cell = gpAdvManager->GetCell(this->x, this->y);
-   if (cell->objType != (LOCATION_TOWN | TILE_HAS_EVENT)) {
-      return this->CalcMobility_orig();
-   } 
-   
-   points = MOVEMENT_POINTS_TERM_CREATURE_MAX;
-   points = (signed __int64)((double)points * gfSSLogisticsMod[GetSSLevel(SECONDARY_SKILL_LOGISTICS)]);
-   if (this->flags & HERO_AT_SEA) {
-      points += 400;
-   }
-   if (this->HasArtifact(ARTIFACT_NOMAD_BOOTS_OF_MOBILITY)) {
-      points += 600;
-   }
-   if (this->HasArtifact(ARTIFACT_TRAVELERS_BOOTS_OF_MOBILITY)) {
-      points += 300;
-   }
-   if (this->HasArtifact(ARTIFACT_TRUE_COMPASS_OF_MOBILITY)) {
-      points += 500;
-   }
-   if (this->ownerIdx >= PLAYER_FIRST 
-	   && this->ownerIdx <= PLAYER_LAST) {
-      if (!gbHumanPlayer[ownerIdx]) {
-         if (gpGame->difficulty >= DIFFICULTY_HARD) {
+  int points;
+  int MOVEMENT_POINTS_TERM_CREATURE_MAX = 1500;
+  int PLAYER_FIRST = 0;
+  int PLAYER_LAST = 5;
+  float gfSSLogisticsMod[] = { 1.0,  1.1,  1.2,  1.3 };
+  
+  mapCell* cell = gpAdvManager->GetCell(this->x, this->y);
+  if (cell != nullptr) {  //This is here because CalcMobility() is called on the first successive start of a New Game in the current running instance of the program
```

Comment:
I think this will be much more readable if you just change this to:

if (cell == nullptr) {
  return this->CalcMobility_orig();
}

....